File: vlc-android/src/org/videolan/vlc/gui/browser/FilePickerFragment.java
/*
 * *************************************************************************
 *  FilePickerFragment.java
 * **************************************************************************
 *  Copyright © 2015 VLC authors and VideoLAN
 *  Author: Geoffrey Métais
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
 *  ***************************************************************************
 */

package org.videolan.vlc.gui.browser;

import android.app.Activity;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.text.TextUtils;
import android.view.View;

import org.videolan.medialibrary.media.MediaLibraryItem;
import org.videolan.medialibrary.media.MediaWrapper;
import org.videolan.vlc.R;
import org.videolan.vlc.util.AndroidDevices;
import org.videolan.vlc.util.FileUtils;
import org.videolan.vlc.util.Strings;
import org.videolan.vlc.viewmodels.browser.BrowserModel;
import org.videolan.vlc.viewmodels.browser.BrowserModelKt;

public class FilePickerFragment extends FileBrowserFragment {

    public static final String EXTRA_MRL = "sub_mrl";

    private static String[] rootDirectories = AndroidDevices.getMediaDirectories();

    @Override
    protected Fragment createFragment() {
        return new FilePickerFragment();
    }

    public boolean isSortEnabled() {
        return false;
    }

    @Override
    public void onCreate(Bundle bundle) {
        final Activity activity = getActivity();
        if (activity != null && activity.getIntent() != null) {
            final Uri uri = getActivity().getIntent().getData();
            if (uri == null || TextUtils.equals(uri.getScheme(), "http")) {
                activity.setIntent(null);
            }
        }
        super.onCreate(bundle);
        setAdapter(new FilePickerAdapter(this));
        setRootDirectory(defineIsRoot());
    }

    @Override
    protected void setupBrowser() {
        viewModel = ViewModelProviders.of(this, new BrowserModel.Factory(getMrl(), BrowserModelKt.TYPE_PICKER, false)).get(BrowserModel.class);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        getBinding().empty.setText(R.string.no_subs_found);
    }

    @Override
    public void onStart() {
        super.onStart();
        getActivity().setTitle(getTitle());
        mSwipeRefreshLayout.setEnabled(false);
    }

    public void onClick(View v, int position, MediaLibraryItem item) {
        final MediaWrapper media = (MediaWrapper) item;
        if (media.getType() == MediaWrapper.TYPE_DIR) browse(media, true);
        else pickFile(media);

    }
    void pickFile(MediaWrapper mw){
        Intent i = new Intent(Intent.ACTION_PICK);
        i.putExtra(EXTRA_MRL, mw.getLocation());
        requireActivity().setResult(Activity.RESULT_OK, i);
        requireActivity().finish();
    }

    public void browseUp() {
        if (isRootDirectory()) requireActivity().finish();
        else if (TextUtils.equals(Strings.removeFileProtocole(getMrl()), AndroidDevices.EXTERNAL_PUBLIC_DIRECTORY)) {
            setMrl(null);
            setRootDirectory(true);
            viewModel.fetch();
        } else if (getMrl() != null) {
            final MediaWrapper mw = new MediaWrapper(Uri.parse(FileUtils.getParent(getMrl())));
            browse(mw, false);
        }
    }

    protected boolean defineIsRoot() {
        if (getMrl() == null) return true;
        if (getMrl().startsWith("file")) {
            final String path = Strings.removeFileProtocole(getMrl());
            for (String directory : rootDirectories) {
                if (path.startsWith(directory))
                    return false;
            }
            return true;
        } else return getMrl().length() < 7;
    }
}

File: vlc-android/src/org/videolan/vlc/gui/browser/FilePickerFragment.kt
/*
 * *************************************************************************
 *  FilePickerFragment.java
 * **************************************************************************
 *  Copyright © 2015 VLC authors and VideoLAN
 *  Author: Geoffrey Métais
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
 *  ***************************************************************************
 */

package org.videolan.vlc.gui.browser

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProviders
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import org.videolan.medialibrary.MLServiceLocator
import org.videolan.medialibrary.interfaces.media.AbstractMediaWrapper
import org.videolan.medialibrary.media.MediaLibraryItem
import org.videolan.vlc.R
import org.videolan.vlc.repository.DirectoryRepository
import org.videolan.vlc.util.AndroidDevices
import org.videolan.vlc.util.FileUtils
import org.videolan.vlc.util.removeFileProtocole
import org.videolan.vlc.viewmodels.browser.BrowserModel
import org.videolan.vlc.viewmodels.browser.TYPE_PICKER

const val EXTRA_MRL = "sub_mrl"
class FilePickerFragment : FileBrowserFragment() {

    override fun createFragment(): Fragment {
        return FilePickerFragment()
    }

    override fun onCreate(bundle: Bundle?) {
        val uri = activity?.intent?.data
        if (uri == null || uri.scheme == "http" || uri.scheme == "content" || uri.scheme == "fd") {
            activity?.intent = null
        }
        super.onCreate(bundle)
        adapter = FilePickerAdapter(this)
    }

    override fun setupBrowser() {
        viewModel = ViewModelProviders.of(this, BrowserModel.Factory(requireContext(), mrl, TYPE_PICKER, false)).get(BrowserModel::class.java)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.empty.setText(R.string.no_subs_found)
    }

    override fun onStart() {
        super.onStart()
        activity?.title = getTitle()
        swipeRefreshLayout.isEnabled = false
    }

    override fun onClick(v: View, position: Int, item: MediaLibraryItem) {
        val media = item as AbstractMediaWrapper
        if (media.type == AbstractMediaWrapper.TYPE_DIR)
            browse(media, true)
        else
            pickFile(media)

    }

    private fun pickFile(mw: AbstractMediaWrapper) {
        val i = Intent(Intent.ACTION_PICK)
        i.putExtra(EXTRA_MRL, mw.location)
        requireActivity().setResult(Activity.RESULT_OK, i)
        requireActivity().finish()
    }

    fun browseUp() {
        when {
            isRootDirectory -> requireActivity().finish()
            TextUtils.equals(mrl?.removeFileProtocole(), AndroidDevices.EXTERNAL_PUBLIC_DIRECTORY) -> {
                mrl = null
                isRootDirectory = true
                viewModel.refresh()
            }
            mrl != null -> {
                val mw = MLServiceLocator.getAbstractMediaWrapper(Uri.parse(FileUtils.getParent(mrl)))
                browse(mw, false)
            }
        }
    }

    override fun defineIsRoot() = mrl?.run {
        if (startsWith("file")) {
            val path = removeFileProtocole()
            val rootDirectories = runBlocking(Dispatchers.IO) { DirectoryRepository.getInstance(requireContext()).getMediaDirectories() }
            for (directory in rootDirectories) if (path.startsWith(directory)) return false
            return true
        } else length < 7
    } ?: true
}
